File: apps/marketplace-api/app/models/marketplace.py
"""
Marketplace-related Pydantic models
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.common import (
    CollaborationOfferingResponse,
    CreatorRequirementsResponse,
)

# ============================================
# LISTING RESPONSE
# ============================================


class ListingMarketplaceResponse(BaseModel):
    """Listing response model for marketplace"""

    id: str
    hotelProfileId: str = Field(alias="hotel_profile_id")
    hotelName: str = Field(alias="hotel_name")
    hotelPicture: str | None = Field(None, alias="hotel_picture")
    ownerEmail: str | None = Field(None, alias="owner_email")
    ownerUserId: str | None = Field(None, alias="owner_user_id")
    name: str
    location: str
    description: str
    accommodationType: str | None = Field(None, alias="accommodation_type")
    images: list[str]
    status: str
    collaborationOfferings: list[CollaborationOfferingResponse] = Field(
        alias="collaboration_offerings"
    )
    creatorRequirements: CreatorRequirementsResponse | None = Field(
        None, alias="creator_requirements"
    )
    createdAt: datetime = Field(alias="created_at")

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


# ============================================
# PLATFORM RESPONSE
# ============================================


class PlatformMarketplaceResponse(BaseModel):
    """Platform response model for marketplace"""

    id: str
    name: str
    handle: str
    followers: int
    engagementRate: float = Field(alias="engagement_rate")
    topCountries: list[dict] | None = Field(None, alias="top_countries")
    topAgeGroups: list[dict] | None = Field(None, alias="top_age_groups")
    genderSplit: dict | None = Field(None, alias="gender_split")

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


# ============================================
# CREATOR RESPONSE
# ============================================


class CreatorMarketplaceResponse(BaseModel):
    """Creator response model for marketplace"""

    id: str
    name: str
    location: str
    shortDescription: str = Field(alias="short_description")
    portfolioLink: str | None = Field(None, alias="portfolio_link")
    profilePicture: str | None = Field(None, alias="profile_picture")
    creatorType: str = Field(alias="creator_type")
    platforms: list[PlatformMarketplaceResponse]
    audienceSize: int = Field(alias="audience_size")
    averageRating: float = Field(alias="average_rating")
    totalReviews: int = Field(alias="total_reviews")
    createdAt: datetime = Field(alias="created_at")

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)

File: apps/marketplace-admin/services/api/marketplace.ts
/**
 * Marketplace API service - fetches public marketplace data
 */
import { apiClient } from './client'

export interface MarketplaceListing {
  id: string
  hotel_profile_id: string
  hotel_name: string
  hotel_picture: string | null
  name: string
  location: string
  description: string
  accommodation_type: string | null
  images: string[]
  status: string
  collaboration_offerings: CollaborationOffering[]
  creator_requirements: CreatorRequirements | null
  created_at: string
}

export interface CollaborationOffering {
  id: string
  listing_id: string
  collaboration_type: 'Free Stay' | 'Paid' | 'Discount'
  availability_months: string[]
  platforms: string[]
  free_stay_min_nights: number | null
  free_stay_max_nights: number | null
  paid_max_amount: number | null
  discount_percentage: number | null
  created_at: string
  updated_at: string
}

export interface CreatorRequirements {
  id: string
  listing_id: string
  platforms: string[]
  min_followers: number | null
  target_countries: string[]
  target_age_min: number | null
  target_age_max: number | null
  target_age_groups: string[] | null
  created_at: string
  updated_at: string
}

export interface MarketplaceCreator {
  id: string
  name: string
  location: string
  short_description: string
  portfolio_link: string | null
  profile_picture: string | null
  platforms: CreatorPlatform[]
  audience_size: number
  average_rating: number
  total_reviews: number
  created_at: string
}

export interface CreatorPlatform {
  id: string
  name: string
  handle: string
  followers: number
  engagement_rate: number
  top_countries: { country: string; percentage: number }[] | null
  top_age_groups: { ageRange: string; percentage: number }[] | null
  gender_split: { male: number; female: number } | null
}

export const marketplaceService = {
  /**
   * Get all marketplace listings (public endpoint)
   */
  getListings: async (): Promise<MarketplaceListing[]> => {
    return apiClient.get<MarketplaceListing[]>('/marketplace/listings')
  },

  /**
   * Get all marketplace creators (public endpoint)
   */
  getCreators: async (): Promise<MarketplaceCreator[]> => {
    return apiClient.get<MarketplaceCreator[]>('/marketplace/creators')
  },
}
